hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/flow/dispatch/logic/RecordFinalizer.java

```diff
@@ -0,0 +1,137 @@
+        switch (dispatch.txnCategory()) {
+            case USER, SCHEDULED -> parentRecordFinalizer.finalizeParentRecord(
+                    dispatch.finalizeContext(),
+                    dispatch.txnInfo().functionality(),
+                    extraRewardReceivers(
+                            dispatch.txnInfo().txBody(), dispatch.txnInfo().functionality(), dispatch.recordBuilder()),
+                    dispatch.handleContext().dispatchPaidRewards());
+            case CHILD, PRECEDING -> childRecordFinalizer.finalizeChildRecord(
+                    dispatch.finalizeContext(), dispatch.txnInfo().functionality());
+        }
+    }
+
+    /**
+     * Returns a set of "extra" account ids that should be considered as eligible for
+     * collecting their accrued staking rewards with the given transaction info and
+     * record builder.
+     *
+     * <p><b>IMPORTANT:</b> Needed only for mono-service fidelity.
+     *
+     * <p>There are three cases, none of which HIP-406 defined as a reward situation;
+     * but were "false positives" in the original mono-service implementation:
+     * <ol>
+     *     <li>For a crypto transfer, any account explicitly listed in the HBAR
+     *     transfer list, even with a zero balance adjustment.</li>
+     *     <li>For a contract operation, any called contract.</li>
+     *     <li>For a contract operation, any account loaded in a child
+     *     transaction (primarily, any account involved in a child
+     *     token transfer).</li>
+     * </ol>
+     *
+     * @param body the {@link TransactionBody} of the transaction
+     * @param function the {@link HederaFunctionality} of the transaction
+     * @param recordBuilder the record builder
+     * @return the set of extra account ids
+     */
+    Set<AccountID> extraRewardReceivers(
+            @Nullable final TransactionBody body,
+            @NonNull final HederaFunctionality function,
+            @NonNull final SingleTransactionRecordBuilderImpl recordBuilder) {
+        if (recordBuilder.status() != SUCCESS || body == null) {
+            return emptySet();
+        }
+        return switch (function) {
+            case CRYPTO_TRANSFER -> zeroAdjustIdsFrom(body.cryptoTransferOrThrow()
+                    .transfersOrElse(TransferList.DEFAULT)
+                    .accountAmounts());
+            case ETHEREUM_TRANSACTION, CONTRACT_CALL, CONTRACT_CREATE -> recordBuilder.explicitRewardSituationIds();
+            default -> emptySet();
+        };
+    }
+
+    /**
+     * Returns any ids from the given list of explicit hbar adjustments that have a zero amount.
+     *
+     * @param explicitHbarAdjustments the list of explicit hbar adjustments
+     * @return the set of account ids that have a zero amount
+     */
+    @NonNull
+    Set<AccountID> zeroAdjustIdsFrom(@NonNull final List<AccountAmount> explicitHbarAdjustments) {
+        Set<AccountID> zeroAdjustmentAccounts = null;
```

Comment:
I think it would be easier to filter a stream here.

Reply:
Yes absolutely. 💯 Fixed